### DISCUSSION TITLE
Multiple Circular Buffers in one sketch?

### DISCUSSION THREAD

**Discussion by WallyWalker** [rlogiacco/CircularBuffer#62] [Q&A] [answered]
Excuse my ignorance, I've been using CircularBuffer with great success for a few simple projects. I now have a case where I'd like to use two separate circular buffers in the same sketch. Is this currently possible to do with CircularBuffer?

**Comment by rlogiacco** [comment]
I don't see a reason why it shouldn't be possible.
Have you tried and encountered any difficulties?

**Comment by WallyWalker** [comment, marked as answer]
I'm confused on the structure and naming. If the following is already used and later called and I want to create a buffer of a different size, do you just have to name the buffer integer something different? 

`CircularBuffer<int, 25> buffer;`

**Reply by digamesystems** [reply]
Hi, guys.

It sounds like you are looking to create two or more circular buffers that have different roles in your application. 

I use a couple in an app I'm working on that use a lidar sensor to monitor traffic. 

Here's the declares for them: 

`
```
const int lidarSamples = 25;
const int lidarHistorySamples = 150;

CircularBuffer<int, lidarSamples> lidarBuffer; // We're going to hang onto the last few points of raw data
                                               //   to visualize what the sensor sees in the moment

CircularBuffer<int, lidarHistorySamples> lidarHistoryBuffer; // A longer buffer for visualization of the history
                                             // before the algorithm makes a decision.
```
`
I just use lidarBuffer and lidarHistoryBuffer independently, without issues. To Wally's comment, I haven't tried to dynamically resize them. I'm not sure the library supports that. (?)

**Reply by rlogiacco** [reply]
You can have as many buffers as your device memory allows for, each one with the size and the data type you like.

No, the library doesn't support buffer size variation: once created a buffer is going to have that specific capacity forever: it's a circular buffer, not a dynamic list.

So, you can have a sketch like:
```cpp
CircularBuffer<int, 25> oneBuffer;
CircularBuffer<char, 250> anotherBuffer;
CircularBuffer<unsigned int, 50> aThirdBuffer;
```

**Reply by WallyWalker** [reply]
> You can have as many buffers as your device memory allows for, each one with the size and the data type you like.
> 
> No, the library doesn't support buffer size variation: once created a buffer is going to have that specific capacity forever: it's a circular buffer, not a dynamic list.
> 
> So, you can have a sketch like:
> 
> ```c++
> CircularBuffer<int, 25> oneBuffer;
> CircularBuffer<char, 250> anotherBuffer;
> CircularBuffer<unsigned int, 50> aThirdBuffer;
> ```

Perfect! This is it, that's the solution. Thanks rlogiacco it's working perfectly in my code to take running averages of analog sensors and in parsing CAN bus info.
